### PR TITLE
update count_wildcard_rule for more scenario

### LINKED ISSUE #5473: Cannot Count(Expr:Wildcard) with DataFrame API
**Describe the bug**

Cannot do `count(Expr::Wildcard)` aggregate with DataFrame API

**To Reproduce**

```rust
    ctx.table("alltypes_plain")
        .await?
        .aggregate(vec![], vec![count(Expr::Wildcard)])?
        .explain(false, false)?
        .show()
        .await?;
```

Throws error:

`Error: Internal("Wildcard expressions are not valid in a logical query plan")`

**Expected behavior**

Should be like when doing SQL:

```rust
    ctx.sql("select count(*) from alltypes_plain")
        .await?
        .explain(false, false)?
        .show()
        .await?;
```

Gives plan:

```
+---------------+---------------------------------------------------+
| plan_type     | plan                                              |
+---------------+---------------------------------------------------+
| logical_plan  | Aggregate: groupBy=[[]], aggr=[[COUNT(UInt8(1))]] |
|               |   TableScan: alltypes_plain projection=[id]       |
| physical_plan | ProjectionExec: expr=[7300 as COUNT(UInt8(1))]    |
|               |   EmptyExec: produce_one_row=true                 |
|               |                                                   |
+---------------+---------------------------------------------------+
```

Where `count(*)` has been transformed to `count(1)` essentially.

**Additional context**

Expansion substitute:

https://github.com/apache/arrow-datafusion/blob/e9852074bacd8c891d84eba38b3417aa16a2d18c/datafusion/expr/src/utils.rs#L43-L45

See SQL substitute function:

https://github.com/apache/arrow-datafusion/blob/e9852074bacd8c891d84eba38b3417aa16a2d18c/datafusion/sql/src/expr/function.rs#L193-L218

### PR DESCRIPTION
**Pull request by jiangzhx**
# Which issue does this PR close?

Closes #5473.

add more scenario

- test_count_wildcard_on_sort
- test_count_wildcard_on_where_in
- test_count_wildcard_on_where_exist
- test_count_wildcard_on_where_scalar_subquery

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/dataframe.rs [hunk 1]
```diff
@@ -32,24 +32,179 @@ use datafusion::error::Result;
...
+    // In the same SessionContext, AliasGenerator will increase subquery_alias id by 1
+    // https://github.com/apache/arrow-datafusion/blame/cf45eb9020092943b96653d70fafb143cc362e19/datafusion/optimizer/src/alias.rs#L40-L43
+    // for compare difference betwwen sql and df logical plan, we need to create a new SessionContext here
+    let ctx = create_join_context()?;
+    let df_results = ctx
+        .table("t1")
+        .await?
+        .filter(Expr::InSubquery {
+            expr: Box::new(col("a")),
+            subquery: Subquery {
```
Comment (alamb) [inline datafusion/core/tests/dataframe.rs:92]: I think you could use `in_subquery` to simplify this test https://docs.rs/datafusion/latest/datafusion/prelude/fn.in_subquery.html

File: datafusion/core/tests/dataframe.rs [hunk 2]
```diff
@@ -32,24 +32,179 @@ use datafusion::error::Result;
...
+    let sql_results = ctx
+        .sql("SELECT a, b FROM t1 WHERE EXISTS (SELECT count(*) FROM t2)")
+        .await?
+        .explain(false, false)?
+        .collect()
+        .await?;
+    let df_results = ctx
+        .table("t1")
+        .await?
+        .filter(Expr::Exists {
```
Comment (alamb) [inline datafusion/core/tests/dataframe.rs:133]: Likewise there  is an `exists` function to help https://docs.rs/datafusion/latest/datafusion/prelude/fn.exists.html

And there are several other places in this PR that this or similar functions can be used when creating subqueries

**Review by alamb — APPROVED** [review]
Looks good to me -- I especially like the test coverage

**Comment by alamb** [comment]
Thank you @jiangzhx

File: datafusion/core/tests/dataframe.rs (hunk 2 above)
Comment (jiangzhx) [reply datafusion/core/tests/dataframe.rs]: thanks,fixed

File: datafusion/core/tests/dataframe.rs (hunk 1 above)
Comment (jiangzhx) [reply datafusion/core/tests/dataframe.rs]: thanks ,fixed

**Review by alamb — APPROVED** [review]
Looks good to me -- thanks again @jiangzhx

**Pull request merged**